examples/dmd/wave_equation.cpp

```diff
@@ -0,0 +1,639 @@
+    dudt_gf.MakeTRef(&fespace, vx, true_offset[0]);
+    u_gf.MakeTRef(&fespace, vx, true_offset[1]);
+
+
+    // 6. Set the initial conditions for u. All boundaries are considered
+    //    natural.
+    FunctionCoefficient u_0(InitialSolution);
+    u_gf.ProjectCoefficient(u_0);
+    Vector u;
+    u_gf.GetTrueDofs(u);
+
+    FunctionCoefficient dudt_0(InitialRate);
+    dudt_gf.ProjectCoefficient(dudt_0);
+    Vector dudt;
+    dudt_gf.GetTrueDofs(dudt);
+
+    // 7. Initialize the conduction operator and the visualization.
+    Array<int> ess_bdr;
+    if (mesh->bdr_attributes.Size())
+    {
+        ess_bdr.SetSize(mesh->bdr_attributes.Max());
+
+        if (dirichlet)
+        {
+            ess_bdr = 1;
+        }
+        else
+        {
+            ess_bdr = 0;
+        }
+    }
+
+    WaveOperator oper(fespace, ess_bdr, speed);
+
+    u_gf.SetFromTrueDofs(u);
+    {
+        ofstream omesh("wave_equation.mesh");
+        omesh.precision(precision);
+        mesh->Print(omesh);
+        ofstream osol("wave_equation-init.gf");
+        osol.precision(precision);
+        u_gf.Save(osol);
+        dudt_gf.Save(osol);
+    }
+
+    VisItDataCollection visit_dc("Wave_Equation", mesh);
+    visit_dc.RegisterField("solution", &u_gf);
+    visit_dc.RegisterField("rate", &dudt_gf);
+    if (visit)
+    {
+        visit_dc.SetCycle(0);
+        visit_dc.SetTime(0.0);
+        visit_dc.Save();
+    }
+
+    socketstream sout;
+    if (visualization)
+    {
+        char vishost[] = "localhost";
+        int  visport   = 19916;
+        sout.open(vishost, visport);
+        if (!sout)
+        {
+            cout << "Unable to connect to GLVis server at "
+                 << vishost << ':' << visport << endl;
+            visualization = false;
+            cout << "GLVis visualization disabled.\n";
+        }
+        else
+        {
+            sout.precision(precision);
+            sout << "solution\n" << *mesh << dudt_gf;
+            sout << flush;
+        }
+    }
+
+    StopWatch fom_timer, dmd_training_timer, dmd_prediction_timer;
+    fom_timer.Start();
+
+    // 8. Perform time-integration (looping over the time iterations, ti, with a
+    //    time-step dt).
+    ode_solver->Init(oper);
+    double t = 0.0;
+
+    vector<double> ts;
+
+    fom_timer.Stop();
+
+    dmd_training_timer.Start();
+
+    int curr_window = 0;
+    vector<CAROM::DMD*> dmd_u;
+    dmd_u.push_back(new CAROM::DMD(u_gf.GetTrueVector().Size(), dt));
+    dmd_u[curr_window]->takeSample(u, t);
+    ts.push_back(t);
+
+
+    dmd_training_timer.Stop();
+
+
+    bool last_step = false;
+    for (int ti = 1; !last_step; ti++)
+    {
+        fom_timer.Start();
+        if (t + dt >= t_final - dt/2)
+        {
+            last_step = true;
+        }
+
+
+        ode_solver->Step(u, dudt, t, dt);
+        fom_timer.Stop();
+
+        dmd_training_timer.Start();
+        u_gf.SetFromTrueDofs(u);
+        dmd_u[curr_window]->takeSample(u_gf.GetTrueVector(), t);
```

Comment:
In MFEM, there are true DOFs and local DOFs on each MPI rank. The local DOFs are all DOFs on the local partition of the mesh, and the true DOFs are a subset of the local DOFs owned by the MPI rank. The local DOFs that are not true DOFs are owned by another MPI rank (these are shared DOFs on the parallel mesh partition interface). In serial, none of this matters, as local and true DOFs are the same.

A `GridFunction` is a `Vector` of local DOFs. The true DOFs `Vector` is given by `u_gf.GetTrueVector()`. For ROM in parallel, it is best to sample only the true DOFs, since the shared DOFs are redundant. The true DOFs vector is also `u` here, which is the `Vector` being integrated in time. There is no need to do `u_gf.SetFromTrueDofs(u);` unless `u_gf` is being used for visualization. So you can simply sample `u` and skip the copying of values between `u` and `u_gf`.

Reply:
This should now be fixed.